biz.aQute.bndlib/src/aQute/bnd/osgi/Analyzer.java

```diff
@@ -2663,7 +2668,12 @@ private boolean analyzeJar(Jar jar, String prefix, boolean okToIncludeDirs, Stri
 						Set<PackageRef> refs = new LinkedHashSet<>(clazz.getReferred());
 						refs.addAll(referencesByAnnotation(clazz));
 						for (PackageRef p : refs) {
-							referred.put(p);
+							Attrs attrs = importPackage.get(p.getFQN());
+							if (attrs != null) {
+								referred.put(p, attrs);
```

Comment:
The `put(PackageRef ref)` method only adds the ref if it is not already there, so it will not overwrite any previous attrs. The `put(PackageRef key, Attrs value)` method blindly overwrites what was already there. Since a given package may be split along the Bundle-Classpath, you end up with only the attrs of the last encountered package with attrs. So if a package exists more than once with attrs along the Bundle-Classpath, last attrs wins. 

So you may want to do

```java
referred.merge(p, false, attrs);
```

instead to merge all encountered attrs. 

This still may not be "right" since if one jar imports without resolution:=optional and another imports with resolution:=optional, the real answer is not resolution:=optional. But merging will end up with resolution:=optional unless the last jar imports with an explicit resolution:=mandatory. So even a blind merge is not correct. You may need to write a method for this which does an explicit merge on a defined set of Import-Package directives and attributes.

Reply:
:+1: